<a href="https://colab.research.google.com/github/marthinx/UdaCity/blob/main/WranglingAndAnalyzeData.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [ ]:
#Import the necessary libraries
import numpy as np
import pandas as pd
import json
import time
import tweepy 
import os
import json
from timeit import default_timer as timer
import re

In [ ]:
# Open file
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [ ]:
import requests
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
image_request = requests.get(url, allow_redirects=True)

open('image_predictions.tsv', 'wb').write(image_request.content)

335079

In [ ]:
image_predictions = pd.read_csv('image-predictions.tsv', sep='\t')


3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [ ]:
import tweepy

consumer_key = 'Njf43Q5O8Dqkm8Eig46RVCOoL'
consumer_secret = 'P2cmdgIM2bu8XDPvPWbd1ZHTX6jgNK6dTouAc6uSMkp0TzkgSz'
access_token = '8446002-xKoJpitX8bT4YgirtbjJjD5U1tuntkwIZAwvIUqTzV'
access_secret = 'FFfLxeeQ2VzCP87zB60jsBoEJNqtrVNgmNwqr6OCkPPm5'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth_handler=auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)


In [ ]:
# os.remove("tweet_json.txt")

# Check if file is already available
if not os.path.isfile('tweet_json.txt'):
# create file and download json data
# run rest of code regardless of downloading
# start the timer 
    start = time.time()
    tweet_ids = archive.tweet_id.values

    tweets_data = []
    tweet_success = []
    tweet_failure = []

    valid = 0
    error = 0


    with open('tweet_json.txt', 'w', encoding='utf-8') as outfile:
        for tweet_id in tweet_ids:

            try:
                data = api.get_status(tweet_id, tweet_mode='extended', 
                                      wait_on_rate_limit = True, 
                                      wait_on_rate_limit_notify = True)
                tweets_data.append(data)
                tweet_success.append(tweet_id)
                print("%s# %s" % (str(i+1), tweet_id))
                json.dump(json_content, outfile)
                outfile.write('\n')
                valid += 1
            except:
                error += 1
                tweet_failure.append(tweet_id)
                print(tweet_id)

    print("Tweets")            
    print("%s %s" % ('Valid:', valid))
    print("%s %s" % ('Errors:', error))

    # stop timer            
    end = time.time()

    print("Time:")
    print(end - start)
else:
  print("tweet_json.txt already generated")

tweet_json.txt already generated


In [ ]:
# BEISPIEL AUS ÜBUNG VERGLEICHEN MIT OBEN
# os.remove("tweet_json.txt")

# Check if file is already available
if not os.path.isfile('tweet_json.txt'):
    # Query Twitter's API for JSON data for each tweet ID in the Twitter archive
    count = 0
    fails_dict = {}
    start = timer()
    # Save each tweet's returned JSON as a new line in a .txt file
    with open('tweet_json.txt', 'w') as outfile:
        # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
        for tweet_id in tweet_ids:
            count += 1
            print(str(count) + ": " + str(tweet_id))
            try:
                tweet = api.get_status(tweet_id, tweet_mode='extended', 
                                      wait_on_rate_limit = True, 
                                      wait_on_rate_limit_notify = True)
                print("Success")
                json.dump(tweet._json, outfile)
                outfile.write('\n')
            except tweepy.TweepError as e:
                print("Fail")
                fails_dict[tweet_id] = e
                pass
    end = timer()
    print(end - start)
    print(fails_dict)
else:
  print("tweet_json.txt already generated")

tweet_json.txt already generated


In [ ]:
df = []
with open('tweet-json.txt', 'r') as file:
    lines = file.readlines()
    for line in lines:
        parsed_json = json.loads(line)
        df.append({'tweet_id': parsed_json['id'],
                        'retweet_count': parsed_json['retweet_count'],
                        'favorite_count': parsed_json['favorite_count']})
        
tweet_json = pd.DataFrame(df, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

In [ ]:
tweet_json.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

In [ ]:
tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null int64
retweet_count     2354 non-null int64
favorite_count    2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


In [ ]:
twitter_archive.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [ ]:
# detect dublicated 'tweet_ids'
sum(twitter_archive['tweet_id'].duplicated())

0

In [ ]:
# analyse 'rating_numerator' and 'rating_denominator' with 'describe()'
twitter_archive[['rating_numerator', 'rating_denominator']].describe()

rating_numerator  rating_denominator
count       2356.000000         2356.000000
mean          13.126486           10.455433
std           45.876648            6.745237
min            0.000000            0.000000
25%           10.000000           10.000000
50%           11.000000           10.000000
75%           12.000000           10.000000
max         1776.000000          170.000000

In [ ]:
# check the values of 'rating_numerator' with 'value_counts()'
twitter_archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [ ]:
# check the values of 'rating_denominator' with 'value_counts()'
twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [ ]:
# check the values of 'source' with 'value_counts()'
twitter_archive.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [ ]:
image_predictions.sample(5)

tweet_id                                          jpg_url  \
1150  732005617171337216  https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg   
372   672984142909456390  https://pbs.twimg.com/media/CVbrcZyVAAA5Wpq.jpg   
736   687102708889812993  https://pbs.twimg.com/media/CYkURJjW8AEamoI.jpg   
637   681261549936340994  https://pbs.twimg.com/media/CXRTw_5WMAAUDVp.jpg   
1661  811744202451197953  https://pbs.twimg.com/media/C0PlCQjXAAA9TIh.jpg   

      img_num               p1   p1_conf  p1_dog                  p2  \
1150        1   English_setter  0.677408    True       Border_collie   
372         1           wombat  0.738780   False              beaver   
736         1     fiddler_crab  0.992069   False               quail   
637         1  Tibetan_terrier  0.382101    True    miniature_poodle   
1661        1         Pekinese  0.386082    True  Labrador_retriever   

       p2_conf  p2_dog                p3   p3_conf  p3_dog  
1150  0.052724    True    cocker_spaniel  0.048572    True  
372   0.133368   False           wallaby  0.032370   False  
736   0.002491   False         rock_crab  0.001513   False  
637   0.095429    True       Maltese_dog  0.065738    True  
1661  0.202862    True  golden_retriever  0.170487    True

In [ ]:
# Analysation of Data 2: prediction image
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [ ]:
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [ ]:
# Analysation of Data 3: tweet_json
tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null int64
retweet_count     2354 non-null int64
favorite_count    2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


In [ ]:
tweet_json.sample(5)

tweet_id  retweet_count  favorite_count
914   756998049151549440           2271            6923
528   808344865868283904          24069           47281
1682  681679526984871937            499            1923
552   804026241225523202          18876           49774
2167  669353438988365824            281             687

In [ ]:
tweet_json.describe()

tweet_id  retweet_count  favorite_count
count  2.354000e+03    2354.000000     2354.000000
mean   7.426978e+17    3164.797366     8080.968564
std    6.852812e+16    5284.770364    11814.771334
min    6.660209e+17       0.000000        0.000000
25%    6.783975e+17     624.500000     1415.000000
50%    7.194596e+17    1473.500000     3603.500000
75%    7.993058e+17    3652.000000    10122.250000
max    8.924206e+17   79515.000000   132810.000000

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



In [ ]:
# Display the twitter_archive table
twitter_archive


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [ ]:
twitter_archive.shape

(2356, 17)

In [ ]:
twitter_archive.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [ ]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [ ]:
sum(twitter_archive['tweet_id'].duplicated())

0

In [ ]:
image_predictions.shape

(2075, 12)

In [ ]:
tweet_json.shape

(2354, 3)

In [ ]:
sum(image_predictions.jpg_url.duplicated())

66

#### General:
##### twitter_archive
- Rows: 2356
- Columns: 17

##### image_predictions
- Rows: 2075
- Columns: 12

##### tweet_json
- Rows: 2354
- Columns: 31


#### Quality 
##### visual:
- 'twitter_archive': missing values in 'in_reply_to_status_id'
- 'twitter_archive': missing values in 'in_reply_to_user_id'
- 'twitter_archive': missing values in 'retweeted_status_id'
- 'twitter_archive': missing values in 'retweeted_status_user_id'
- 'twitter_archive': missing values in 'retweeted_status_timestamp'
- 'twitter_archive': missing values in 'doggo'
- 'twitter_archive': missing values in 'floofer'
- 'twitter_archive': missing values in 'pupper'
- 'twitter_archive': missing values in 'puppo'
- 'twitter_archive': source is a url and not the direct name of the source
- 'twitter_archive': 'doggo', 'floofer', 'pupper' and 'puppo' should be combined to a new column 'type'   
- 'twitter_archive': 'rating_numerator' and 'rating_denominator' some values are not correct. Should be between 1 - 12


#### programmatic
- 'image_prediction': drop '66' dublicated entries

In [ ]:
twitter_counts_df.sample(5)


NameError: name 'twitter_counts_df' is not defined

In [ ]:
twitter_counts_df['tweet_id'] = tweet_success
twitter_counts_df = twitter_counts_df[['tweet_id', 'favorite_count', 'retweet_count']]
twitter_counts_df.head()

In [ ]:
twitter_counts_df.describe()

### Quality issues
The four main data quality dimensions are:

Completeness: missing data?
Validity: does the data make sense?
Accuracy: inaccurate data? (wrong data can still show up as valid)
Consistency: standardization?

Here are my findings:

1. 

2. Delete all not needed coloums

3. Remove all replies and retweeds datasets

4. Source is a url and not the direct name of the source. It is not good readable

5. 'doggo', 'floofer', 'pupper' and 'puppo' should be combined to a new column 'type'

6. 'rating_numerator'  some values are not correct. The values should be between 1 - 12

7. 'rating_denominator' some values are not correct. Should be between 1 - 10

8. we only need one table -> merge Datasets

### Tidiness issues
Three requirements for tidiness:

Each variable forms a column
Each observation forms a row
Each type of observational unit forms a table

Here are my findings:

1. twitter_archive: 'doggo', 'floofer', 'pupper' and 'puppo' should be combined to a new column 'type' 

2. the tables should be combined because all the data can be used  

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [ ]:
# Make copies of original pieces of data
twitter_archive_clean = twitter_archive.copy()
image_predictions_clean = image_predictions.copy()
tweet_json_clean = tweet_json.copy()

### Issue #1:

#### Define:
we only need one table -> merge Datasets

#### Code

#### Test

In [ ]:
twitter_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

### Issue #2: Delete all not needed coloums


### Issue #3: twitter_archive: Remove all replies and retweeds datasets

In [ ]:
# remove replies
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.in_reply_to_status_id.isnull()]

# remove retweets
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.retweeted_status_id.isnull()]

In [ ]:
twitter_archive_clean['rating_numerator'].value_counts()

12      486
10      436
11      413
13      287
9       153
8        98
7        52
14       38
5        34
6        32
3        19
4        16
2         9
1         5
26        1
44        1
165       1
24        1
60        1
50        1
144       1
80        1
84        1
88        1
121       1
204       1
420       1
1776      1
27        1
45        1
75        1
99        1
0         1
Name: rating_numerator, dtype: int64

In [ ]:
# delete duplicated 'jpg_url' entries
image_predictions_clean = image_predictions_clean.drop_duplicates(subset=['jpg_url'], keep='last')

# and check
sum(image_predictions_clean.jpg_url.duplicated())

0

### Issue #4:

#### Define:
remove all the html header informations in 'source' and replace it only with the label text in the value

#### Code

#### Test

In [ ]:
twitter_archive_clean.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     1964
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       31
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [ ]:
# twitter_archive_clean['source_name'] = twitter_archive_clean['source'].apply(lambda x: re.split('https:\/\/.*', str(x))[0])

In [ ]:
for text in twitter_archive_clean['source']:
    label = re.search('>(.+?)<',text)
    if label: 
        twitter_archive_clean['source_name'] = label.group(1)
        # print(twitter_archive_clean['source_name'])
    

In [ ]:
twitter_archive_clean.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo         source_name  
0                  10   Phineas  None    None   None  None  Twitter for iPhone  
1                  10     Tilly  None    None   None  None  Twitter for iPhone  
2                  10    Archie  None    None   None  None  Twitter for iPhone  
3                  10     Darla  None    None   None  None  Twitter for iPhone  
4                  10  Franklin  None    None   None  None  Twitter for iPhone

In [ ]:
twitter_archive_clean.source_name.value_counts()

Twitter for iPhone    2097
Name: source_name, dtype: int64

In [ ]:
#check
twitter_archive_clean.source_name.value_counts()

Twitter for iPhone    2097
Name: source_name, dtype: int64

In [ ]:
twitter_archive_clean = twitter_archive_clean.drop(columns='source')


#### Test

### Issue #5:

#### Define
'doggo', 'floofer', 'pupper' and 'puppo' should be combined to a new column 'type'

#### Code

In [ ]:
twitter_archive_clean[[ 'doggo', 'floofer', 'pupper', 'puppo']].sample(5)

doggo floofer  pupper puppo
1385  None    None    None  None
2267  None    None    None  None
458   None    None    None  None
352   None    None  pupper  None
1233  None    None    None  None

In [ ]:
twitter_archive_clean['dog_type'] = twitter_archive_clean['text'].str.extract('(doggo|floofer|pupper|puppo)')


#### Test

In [ ]:
twitter_archive_clean[[ 'dog_type', 'doggo', 'floofer', 'pupper', 'puppo']].sample(100)

dog_type  doggo floofer  pupper  puppo
941       NaN   None    None    None   None
2250      NaN   None    None    None   None
1068      NaN   None    None    None   None
1130      NaN   None    None    None   None
1592      NaN   None    None    None   None
1581      NaN   None    None    None   None
1113    doggo  doggo    None  pupper   None
1937   pupper   None    None  pupper   None
1115      NaN   None    None    None   None
129     puppo   None    None    None  puppo
1563   pupper   None    None  pupper   None
1795      NaN   None    None    None   None
1341      NaN   None    None    None   None
1993      NaN   None    None    None   None
1778      NaN   None    None    None   None
104       NaN   None    None    None   None
1290      NaN   None    None    None   None
2345      NaN   None    None    None   None
1443      NaN   None    None    None   None
775       NaN   None    None    None   None
1324      NaN   None    None    None   None
637       NaN   None    None    None   None
1857      NaN   None    None    None   None
1044      NaN   None    None    None   None
1907   pupper   None    None  pupper   None
1489   pupper   None    None  pupper   None
2108      NaN   None    None    None   None
1424      NaN   None    None    None   None
903       NaN   None    None    None   None
1820      NaN   None    None    None   None
...       ...    ...     ...     ...    ...
1739      NaN   None    None    None   None
2083      NaN   None    None    None   None
164       NaN   None    None    None   None
1868      NaN   None    None    None   None
2009   pupper   None    None  pupper   None
2145      NaN   None    None    None   None
2102      NaN   None    None    None   None
2081      NaN   None    None    None   None
1510      NaN   None    None    None   None
718       NaN   None    None    None   None
1072      NaN   None    None    None   None
1165      NaN   None    None    None   None
38        NaN   None    None    None   None
347       NaN   None    None    None   None
1487   pupper   None    None  pupper   None
968       NaN   None    None    None   None
1933      NaN   None    None    None   None
304    pupper   None    None  pupper   None
2279      NaN   None    None    None   None
1807   pupper   None    None  pupper   None
1303      NaN   None    None    None   None
1967   pupper   None    None  pupper   None
1645      NaN   None    None    None   None
756       NaN   None    None    None   None
1468      NaN   None    None    None   None
651       NaN   None    None    None   None
1400      NaN   None    None    None   None
1134      NaN   None    None    None   None
2080      NaN   None    None    None   None
2271      NaN   None    None    None   None

[100 rows x 5 columns]

### Issue #8: 

#### Define
we only need one table -> merge Datasets

#### Code

In [ ]:
twitter_clean = pd.concat([twitter_archive_clean, image_predictions_clean, tweet_json_clean], join='outer', axis=1)

#### Test

In [ ]:
twitter_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

In [ ]:
twitter_dogs.to_csv('twitter_archive_master.csv', encoding='utf-8', index=False)

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

In [ ]:
twitter_dogs['dog_type'].value_counts()

### Insights:
1. Which dogs breed has the highest rating  ( rating_numerator )?

2.Which dogs breed has the highest rating  ( rating_denominator )?


3.

### Visualization

#### 1. Which dogs breed has the highest rating ( rating_numerator )?

In [ ]:
df_final.filter(['predection1','rating_numerator']).dropna().groupby(['predection1']).max().sort_values(by = 'rating_numerator').tail(10).plot(kind= 'barh', color = 'm', title= 'Top 10 dogs breed have max rating numerator' )

#### 2. Which dogs breed has the highest rating ( rating_denominator )?

In [ ]:
df_final.filter(['predection1','rating_denominator']).dropna().groupby(['predection1']).max().sort_values(by = 'rating_denominator').tail(10).plot(kind='barh',
color="m", title= 'Top 10 dogs breed have max rating denominator' )

In [ ]:
sns.lmplot(x="retweet_count", 
           y="favorite_count", 
           data=twitter_dogs,
           size = 5,
           aspect=1.3,
           scatter_kws={'alpha':1/5});

plt.title('Favorite Count vs. Retweet Count');
plt.xlabel('Retweet Count');
plt.ylabel('Favorite Count');